File: JavaScript/0005_Longest_Palindromic_Substring.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
    let startIndex = 0;             // start index of palindrome
    let endIndex = 0;               // end index of palindrome

    if(s.length < 2) {
        return s;
    }
    
    // from all elements in s, expand mirrored area to find the longest palindrome
    for (let center1 = 0; center1 < s.length; center1++) {
        // consider both odd and even palindromes
        for (let center2 of [center1, center1+1]) {
            // center1 == center2 for odd palindrome, center1+1 == center2 for even palindrome
            let startTemp = center1;
            let endTemp = center2;
            // start index >= 0 && end index <= length of s && start and end elements are identical
            while(startTemp >= 0 && endTemp <= s.length && s[startTemp] === s[endTemp]) {
                // find the longest palindrome index
                [startIndex, endIndex] = (endTemp-startTemp+1) > (endIndex-startIndex+1) ? [startTemp, endTemp] : [startIndex, endIndex];
                // expand mirrored area
                startTemp--, endTemp++;
            }
        }
    }

    return s.substring(startIndex, endIndex+1);
};

File: Python/0005_Longest_Palindromic_Substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        
        # if length of s is less than 0, there is no palindrome
        if len(s) < 1:
            return ""
        # if length of s is 1, s is palindrome itself
        if len(s) == 1:
            return s
        
        maxLen = 1
        start = 0

        # from all elements in s, find the longest palindrome
        for end in range(len(s)):
            # if s[end-maxLen-1:end+1] is palindrome, it means length of palindrome is increased by 2
            if end-maxLen >=1 and s[end-maxLen-1:end+1]==s[end-maxLen-1:end+1][::-1]:
                start = end-maxLen-1
                maxLen+=2
                continue
            # if s[end-maxLen:end+1] is palindrome, it means length of palindrome is increased by 1
            if end-maxLen >= 0 and s[end-maxLen:end+1]==s[end-maxLen:end+1][::-1]:
                start=end-maxLen
                maxLen+=1

        return s[start:start+maxLen]
